// @ethereumjs/evm/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateEOF = exports.RustBN254 = exports.NobleBN254 = exports.NobleBLS = exports.Message = exports.MCLBLS = exports.getOpcodesForHF = exports.getActivePrecompiles = exports.EVMMockBlockchain = exports.EVMError = exports.EVM = exports.EOFContainer = void 0;
const container_ts_1 = require("./eof/container.js");
Object.defineProperty(exports, "EOFContainer", { enumerable: true, get: function () { return container_ts_1.EOFContainer; } });
Object.defineProperty(exports, "validateEOF", { enumerable: true, get: function () { return container_ts_1.validateEOF; } });
const errors_ts_1 = require("./errors.js");
Object.defineProperty(exports, "EVMError", { enumerable: true, get: function () { return errors_ts_1.EVMError; } });
const evm_ts_1 = require("./evm.js");
Object.defineProperty(exports, "EVM", { enumerable: true, get: function () { return evm_ts_1.EVM; } });
const message_ts_1 = require("./message.js");
Object.defineProperty(exports, "Message", { enumerable: true, get: function () { return message_ts_1.Message; } });
const index_ts_1 = require("./opcodes/index.js");
Object.defineProperty(exports, "getOpcodesForHF", { enumerable: true, get: function () { return index_ts_1.getOpcodesForHF; } });
const index_ts_2 = require("./precompiles/index.js");
Object.defineProperty(exports, "MCLBLS", { enumerable: true, get: function () { return index_ts_2.MCLBLS; } });
Object.defineProperty(exports, "NobleBLS", { enumerable: true, get: function () { return index_ts_2.NobleBLS; } });
Object.defineProperty(exports, "NobleBN254", { enumerable: true, get: function () { return index_ts_2.NobleBN254; } });
Object.defineProperty(exports, "RustBN254", { enumerable: true, get: function () { return index_ts_2.RustBN254; } });
Object.defineProperty(exports, "getActivePrecompiles", { enumerable: true, get: function () { return index_ts_2.getActivePrecompiles; } });
const types_ts_1 = require("./types.js");
Object.defineProperty(exports, "EVMMockBlockchain", { enumerable: true, get: function () { return types_ts_1.EVMMockBlockchain; } });
__exportStar(require("./logger.js"), exports);
__exportStar(require("./binaryTreeAccessWitness.js"), exports);
__exportStar(require("./constructors.js"), exports);
__exportStar(require("./params.js"), exports);
